File: USB/mcc-libusb/test-usb3100.c
/*
 *  Copyright (c) 2015 Warren J. Jasper <wjasper@ncsu.edu>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>
#include <stdint.h>

#include "pmd.h"
#include "usb-3100.h"

#define MAX_STR 255

/* Test Program */
int toContinue()
{
  int answer;
  answer = 0; //answer = getchar();
  printf("Continue [yY]? ");
  while((answer = getchar()) == '\0' ||
    answer == '\n');
  return ( answer == 'y' || answer == 'Y');
}

int main(int argc, char **argv) {

  hid_device *hid = 0x0;
  float volts;
  int flag;
  uint8_t channel, range;
  int temp, i;
  int ch;
  uint16_t value;
  wchar_t serial[64];
  wchar_t wstr[MAX_STR];

  int ret;
  uint8_t memory[62];

  ret = hid_init();
  if (ret < 0) {
    fprintf(stderr, "hid_init failed with return code %d\n", ret);
    return -1;
  }

  if ((hid = hid_open(MCC_VID, USB3101_PID, NULL)) > 0) {
    printf("USB 3101 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3102_PID, NULL)) > 0) {
    printf("USB 3102 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3103_PID, NULL)) > 0) {
    printf("USB 3103 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3104_PID, NULL)) > 0) {
    printf("USB 3104 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3105_PID, NULL)) > 0) {
    printf("USB 3105 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3106_PID, NULL)) > 0) {
    printf("USB 3106 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3110_PID, NULL)) > 0) {
    printf("USB 3110 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3112_PID, NULL)) > 0) {
    printf("USB 3112 Device is found!\n");
  } else if ((hid = hid_open(MCC_VID, USB3114_PID, NULL)) > 0) {
    printf("USB 3114 Device is found!\n");
  } else {
    fprintf(stderr, "USB 31XX  not found.\n");
    exit(1);	
  }

  /* config mask 0x01 means all inputs */
  usbDConfigPort_USB31XX(hid, DIO_DIR_OUT);
  usbDOut_USB31XX(hid, 0);

  // Configure all analog channels for 0-10V output
  for (i = 0; i < 8; i++) {
    usbAOutConfig_USB31XX(hid, i, UP_10_00V);
  }

    while(1) {
    printf("\nUSB 31XX Testing\n");
    printf("----------------\n");
    printf("Hit 'a' for analog output of different levels.\n");
    printf("Hit 'o' for analog output.\n");
    printf("Hit 'b' to blink \n");
    printf("Hit 'c' to test counter\n");
    printf("Hit 'd' to test digital output\n");
    printf("Hit 'i' to test digital input\n");
    printf("Hit 'I' for information\n");
    printf("Hit 'e' to exit\n");
    printf("Hit 'g' to get serial number\n");
    printf("Hit 'r' to reset\n");
    printf("Hit 's' to get status\n");
    printf("Hit 'R' to read memory\n");

    while((ch = getchar()) == '\0' ||
      ch == '\n');

    switch(ch) {
      case 'a':
	printf("Testing the analog output...\n");
        printf("Enter channel [0-15]: ");
        scanf("%d", &temp);
        channel = (uint8_t) temp;
	for (value = 0; value < 0xfff0; value += 0xf) {
	  usbAOut_USB31XX(hid, channel, (uint16_t) value, 0);
	}
        usbAOut_USB31XX(hid, channel, 0x0, 0);
        break;
      case 'o':
        printf("Testing the analog output for a single channel.\n");
	printf("Enter channel [0-15]: ");
        scanf("%d", &temp);
        channel = (uint8_t) temp;
       	printf("Enter a range: 0 = 0-10V, 1 = +/- 10V, 2 = 0-20mA ");
	scanf("%d", &temp);
        range = (uint8_t) temp;
        printf("Enter a voltage: ");
        scanf("%f", &volts);
        value = volts_USB31XX(range, volts);
        usbAOutConfig_USB31XX(hid, channel, range);
        usbAOut_USB31XX(hid, channel, value, 0);
        break;
      case 'b': /* test to see if led blinks  4 times*/
        usbBlink_USB31XX(hid, 4);
        break;
      case 'c':
        printf("connect CTR and DIO0\n");
        usbInitCounter_USB31XX(hid);
        sleep(1);
        flag = fcntl(fileno(stdin), F_GETFL);
        fcntl(0, F_SETFL, flag | O_NONBLOCK);
        do {
          usbDOut_USB31XX(hid, 1);
	  usleep(200000);
          usbDOut_USB31XX(hid, 0);
	  printf("Counter = %d\n",usbReadCounter_USB31XX(hid));
        } while (!isalpha(getchar()));
        fcntl(fileno(stdin), F_SETFL, flag);
        break;
      case 'd':
	printf("\nTesting Digital I/O....\n");
        printf("Enter a byte number [0-0xff]: " );
        scanf("%x", &temp);
        usbDConfigPort_USB31XX(hid, DIO_DIR_OUT);
        usbDOut_USB31XX(hid, (uint8_t)temp);
        break;
      case 'i':
      	printf("\nTesting Digital Input....\n");
	usbDConfigPort_USB31XX(hid, DIO_DIR_IN);
	usbDIn_USB31XX(hid, (uint8_t*) &temp);
	temp &= 0xff;
	printf("Digital Input = %#x\n", temp);
	break;
      case 'I':
        // Read the Manufacuter String
        ret = hid_get_manufacturer_string(hid, wstr, MAX_STR);
        printf("Manufacturer String: %ls\n", wstr);
        // Read the Product String
        ret = hid_get_product_string(hid, wstr, MAX_STR);
        printf("Product String: %ls\n", wstr);
        // Read the Serial Number String
        ret = hid_get_serial_number_string(hid, wstr, MAX_STR);
        printf("Serial Number String: %ls\n", wstr);
        break;            
      case 's':
        printf("Status = %#x\n", usbGetStatus_USB31XX(hid));
	break;
      case 'r':
        usbReset_USB31XX(hid);
        return 0;
	break;
      case 'e':
        hid_close(hid);
        hid_exit();
        return 0;
	break;
      case 'g':
        hid_get_serial_number_string(hid, serial, 64);
        printf("Serial Number = %ls\n", serial);
        break;
      case 'R':
        memset(memory, 0x0, 62);
        usbReadMemory_USB31XX(hid, 0x0000, 60, memory);
        printf("reading from EEPROM: \n");
	for (i = 0; i < 62; i+=2) {
	  printf("address = %#x \t value = %#x \t\t", i, memory[i]);
  	  printf("address = %#x \t value = %#x \t\n", i+1, memory[i+2]);
	}
	memset(memory, 0x0, 62);
        usbReadMemory_USB31XX(hid, 0x0100, 62, memory);
        printf("\nreading from FLASH: \n");
	for (i = 0; i < 62; i+=2) {
	  printf("address = %#x \t value = %#x \t\t", i+0x100, memory[i]);
  	  printf("address = %#x \t value = %#x \t\n", i+0x101, memory[i+2]);
	}
        break;
      default:
        break;
    }
  }
}

File: python/linux_drivers/usb/test-usb3100.py
#! /usr/bin/python3
#
# Copyright (c) 2018 Warren J. Jasper <wjasper@ncsu.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from usb_3100 import *
import time
import sys
import fcntl
import os
import math

def toContinue():
  answer = input('Continue [yY]? ')
  if (answer == 'y' or answer == 'Y'):
    return True
  else:
    return False

def main():
  # initalize the class
  try:
    usb3100 = usb_3101()
  except:
    try:
      usb3100 = usb_3102()
    except:
      try:
        usb3100 = usb_3103()
      except:
        try:
          usb3100 = usb_3104()
        except:
          try:
            usb3100 = usb_3105()
          except:
            try:
              usb3100 = usb_3106()
            except:
              try:
                usb3100 = usb_3110()
              except:
                try:
                  usb3100 = usb_3112()
                except:
                  try:
                    usb3100 = usb_3114()
                  except:
                    print('No USB-31XX device found')
                    return

  print("Manufacturer: %s" % usb3100.h.get_manufacturer_string())
  print("Product: %s" % usb3100.h.get_product_string())
  print("Serial No: %s" % usb3100.h.get_serial_number_string())

  # config mask 0x01 means all inputs
  usb3100.DConfigPort(usb3100.DIO_DIR_OUT)
  usb3100.DOut(0)

  # Configure all analog channels for 0-10V output
  for i in range(8):
    usb3100.AOutConfig(i, usb3100.UP_10_00V)
    print('channel = ', i, 'slope = ', format(usb3100.CalTable[i].slope,'.4f'), 'intercept = ',format(usb3100.CalTable[i].intercept,'.4f'))

  while True:
    print("\nUSB 31XX Testing")
    print("----------------");
    print("Hit 'a' for analog output of different levels.")
    print("Hit 'o' for analog output.")
    print("Hit 'b' to blink")
    print("Hit 'c' to test counter")
    print("Hit 'd' to test digital output")
    print("Hit 'i' to test digital input")
    print("Hit 'I' for information")
    print("Hit 'e' to exit")
    print("Hit 'g' to get serial number")
    print("Hit 'r' to reset")
    print("Hit 's' to get status")
    print("Hit 'R' to read memory")

    ch = input('\n')
    if ch == 'b':
      count = int(input('Enter number of times to blink: '))
      usb3100.Blink(count)
    elif ch == 'a':
      print('Testing the analog output ...')
      chan = int(input('Enter channel [0-15]: '))
      for value in range(0,0xfff0,0xf):
        usb3100.AOut(channel, value, 0)
      usb3100.AOut(channel, 0x0, 0)
    elif ch == 'o':
      print('Testing the analog output for a single channel.')
      channel = int(input('Enter channel [0-15]: '))
      gain = int(input('Enter a range: 0 = 0-10V, 1 = +/- 10V, 2 = 0-20mA'))
      voltage = float(input('Enter a voltage: '))
      value = usb3100.volts(gain, voltage)
      usb3100.AOutConfig(channel, gain)
      usb3100.AOut(channel, value, 0)
    elif ch == 'c':
      print('Connect CTR and DIO0')
      print('Hit s <CR> to stop.')
      usb3100.CInit()
      usb3100.DConfigPort(usb3100.DIO_DIR_OUT)
      time.sleep(1.)
      flag = fcntl.fcntl(sys.stdin, fcntl.F_GETFL)
      fcntl.fcntl(sys.stdin, fcntl.F_SETFL, flag|os.O_NONBLOCK)
      while True:
        usb3100.DOut(0x1)
        time.sleep(0.2)
        usb3100.DOut(0x0)
        time.sleep(0.2)
        print('Counter = ', usb3100.CIn())
        c = sys.stdin.readlines()
        if (len(c) != 0):
          fcntl.fcntl(sys.stdin, fcntl.F_SETFL, flag)
          break
    elif ch == 'd':
      print('Testing Digital I/O ...')
      value = hex(input('Enter a byte value [0-0xff]: '))
      usb3100.DConfigPort(usb3100.DIO_DIR_OUT)
      usb3100.DOut(value & 0xff)
    elif ch == 'i':
      print('Testing Digital Input ...')
      usb3100.DConfigPort(usb3100.DIO_DIR_IN)
      value = usb3100.DIn()
      print('Digital Input =', hex(value))
    elif ch == 'e':
      usb3100.h.close()
      exit(0)
    elif ch == 'I':
      print("Manufacturer: %s" % usb3100.h.get_manufacturer_string())
      print("Product: %s" % usb3100.h.get_product_string())
      print("Serial No: %s" % usb3100.h.get_serial_number_string())
    elif ch == 'g':
      print("Serial No: %s" % usb3100.h.get_serial_number_string())
    elif ch == 'r':
      usb3100.Reset()
      return 0
    elif ch == 'R':
      memory = usb3100.MemRead(0x0000, 62)
      print('reading from EEPROM: ')
      for i in range(0,60,2):
        print('address =', hex(i), '\t\tvalue = ', hex(memory[i]),'\taddress =', hex(i+1), '\t\tvalue = ', hex(memory[i+1]))
      print(' ')
      memory = usb3100.MemRead(0x0100, 62)
      for i in range(0,62,2):
        print('address =', hex(i+0x100), '\tvalue = ', hex(memory[i]),'\taddress =', hex(i+0x101), '\tvalue = ', hex(memory[i+1]))
    elif ch == 's':
      print('Status = ',hex(usb3100.Status()))
      
if __name__ == "__main__":
  main()
